examples/opmpack.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+  Copyright 2018 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+#include <getopt.h>
+#include <boost/filesystem.hpp>
+
+#include <opm/parser/eclipse/Parser/Parser.hpp>
+#include <opm/parser/eclipse/Parser/ParseContext.hpp>
+#include <opm/parser/eclipse/Parser/InputErrorAction.hpp>
+#include <opm/parser/eclipse/Deck/Deck.hpp>
+
+
+inline void pack_deck( const char * deck_file, std::ostream& os) {
+    Opm::ParseContext parseContext(Opm::InputError::WARN);
+    Opm::Parser parser;
+
+    auto deck = parser.parseFile(deck_file, parseContext);
+    os << deck;
+
+}
+
+
+void print_help_and_exit() {
+    const char * help_text = R"(
+The opmpack program will load a deck, resolve all include
+files and then print it out again on stdout. All comments
+will be stripped and the value types will be validated.
+
+By passing the option -o you can redirect the output to a file
+or a directory.
+
+Print on stdout:
+
+   opmpack  /path/to/case/CASE.DATA
+
+
+Print MY_CASE.DATA in /tmp:
+
+    opmpack /path/to/MY_CASE.DATA -o /tmp
```

Comment:
This (and the next) example breaks on BSD (and therefore also macos), since getopt() behaves different there (it does not mutate argv). From Wikipedia:

> getopt is a system dependent function. The implementation of getopt in GNU C Library does permute the contents of the argument vector as it scans, so that eventually all the non-option arguments are at the end. On the contrary, the implementation of getopt in BSD C Library does not permute the argument vector and returns -1 if it encounters a non-option argument.

As written, the program works fine also on BSD if any options (i.e. the `-o <output>` part) precede the non-option argument (data filename). I suggest that you rewrite this help text to use this order (the other order would work on Linux as a bonus). The cumbersome alternative is to use only POSIX-guaranteed getopt() features and still provide the order flexibility, which would require rewriting the program.

Reply:
Help text updated